Creating a custom Page block attribute type (#586)
ezsystems/developer-documentation: docs/guide/extending_page.md

            # ...
            visible: false
        ```
    

>     #### Creating blocks with custom attributes

Reviewer: Maybe "Custom block attributes" to keep the heading shorter

    
    You can create Page blocks with custom attributes.

>     The following steps demonstrate the process of creating a text custom attribute type.

Reviewer: ```suggestion
```
The line is not needed (what the section is about should be self-evident from the heading)

    
    First, define the attribute type.

>     You can use one of the types available from `EzPlatformPageFieldtypeBundle` in `src/lib/Form/Type/BlockAttribute/*`.

Reviewer: ```suggestion
    You can use one of the types available in `ezplatform-page-fieldtype/src/lib/Form/Type/BlockAttribute/*`.
```
I think providing path this way is easier for the user to find

>     You can also use one of the [built-in Symfony types](https://symfony.com/doc/current/forms.html#built-in-field-types), eg. `AbstractType` for any custom type or `IntegerType` for numeric types.

Reviewer: ```suggestion
    You can also use one of the [built-in Symfony types](https://symfony.com/doc/current/forms.html#built-in-field-types), e.g. `AbstractType` for any custom type or `IntegerType` for numeric types.
```

    
    To define the type, create a file `src/AppBundle/Block/Attribute/MyStringAttributeType.php` that contains:
    
    ``` php hl_lines="5 6 15"
    <?php declare(strict_types=1);
    
    namespace AppBundle\Block\Attribute;
    
    use Symfony\Component\Form\AbstractType;
    use Symfony\Component\Form\Extension\Core\Type\TextType;
    
    class MyStringAttributeType extends AbstractType
    {
        public function getParent()
        {
            return TextType::class;
        }
    
        public function getBlockPrefix()
        {
            return 'my_string_attribute';
        }
    }
    ```
    
    Note that the attribute uses `AbstractType` (line 5) and `TextType` (line 6).
    Adding `getBlockPrefix` (line 15) returns a unique prefix key for a custom template of the attribute.
    
    Next, configure a template by creating a `src/AppBundle/Resources/views/custom_form_templates.html.twig` file:
    

>     ``` twig

Reviewer: ```suggestion
    ``` html+twig
```

    {% block my_string_attribute_widget %}
        <div style='background: red'>
            <h2>My String</h2>
            {{ form_widget(form) }}
        </div>
    {% endblock %}
    
    {# more templates here #}
    ```
    
    Add the template to `Resources/config/ez_field_templates.yml`:
    
    ``` yaml
    system:
        default:
            field_templates:
                - { template: 'AppBundle:ezform_field.html.twig', priority: 0 }
    ```
    
    At this point, the attribute type configuration is complete, but it requires a mapper.
    Depending on the complexity of the type, you can use a `GenericFormTypeMapper` or create your own.
    
    For a generic mapper, register it as a service by adding:
    
    ``` yaml
     my_application.block.attribute.my_string:
            class: EzSystems\EzPlatformPageFieldType\FieldType\Page\Block\Attribute\FormTypeMapper\GenericFormTypeMapper
            arguments:
                $formTypeClass: AppBundle\Block\Attribute\MyStringAttributeType
            tags:
                - { name: ezplatform.page_builder.attribute_form_type_mapper, alias: my_string }
    ```
    

>     For creating your own mapper, create a class that inherits after `EzSystems\EzPlatformPageFieldType\FieldType\Page\Block\Attribute\FormTypeMapper\AttributeFormTypeMapperInterface`.

Reviewer: ```suggestion
    For creating your own mapper, create a class that inherits from `EzSystems\EzPlatformPageFieldType\FieldType\Page\Block\Attribute\FormTypeMapper\AttributeFormTypeMapperInterface`.
```

Or, actually, shouldn't it be "implements interface"?

>     Then, register the class along with a tag by creating a `src/AppBudle/Block/Attribute/MyStringAttributeMapper.php` file:

Reviewer: ```suggestion
    Then, register the class along with a tag by creating a `src/AppBundle/Block/Attribute/MyStringAttributeMapper.php` file:
```

         * @param \EzSystems\EzPlatformPageFieldType\FieldType\Page\Block\Definition\BlockAttributeDefinition $blockAttributeDefinition
         * @param array $constraints
         *
         * @return \Symfony\Component\Form\FormBuilderInterface
         * @throws \Exception
         */
        public function map(
            FormBuilderInterface $formBuilder,
            BlockDefinition $blockDefinition,
            BlockAttributeDefinition $blockAttributeDefinition,
            array $constraints = []
        ): FormBuilderInterface {
            return $formBuilder->create(
                'value',
                MyStringAttributeType::class,
                [
                    'constraints' => $constraints,
                ]
            );
        }
    }
    ```
    
    The final step is to register your mapper as a service:
    
    ``` yaml
    AppBundle\Block\Attribute\MyStringAttributeMapper:
            tags:
                - { name: ezplatform.page_builder.attribute_form_type_mapper, alias: my_string }
    ```
    
    Now, create a block containing your custom attribute:
    
    ``` yaml hl_lines="8 9 10 11"

>     blocks:

Reviewer: Let's add `ezplatform_page_fieldtype:` before, to be consistent with earlier examples
